### PR TITLE
Use class properties + UXBase class

### PR DESCRIPTION
**Pull request by sadafie**
Not too much to gain here, but good to have.

### CODE REVIEW

**Review by sergeychernyshev — COMMENTED** [review]
It feels strange to define a property and then set it in constructor.

Alternative is to set defaults using class property and those that require calculation in constructor, use in constructor. But this creates two places to define variable which I am not a huge fan of either.

Are there best practices already for this, @mmcgahan?

**Comment by mmcgahan** [comment]
yeah, neither option feels ideal, although we could double down on adopting patterns from React and store the `config`/props object in `this.props` the same way `React.Component` works, and then use class properties to assign values based on `this.props` - we wouldn't need any constructors in that case, which IMO is valuable

freehand implementation:

```js
class UXBase {
  constructor(props) {
    this.props = props;
  }
}

class ExpectedMark extends UXBase {
  // ExpectedMark no longer needs constructor - use `this.props.name` instead of `this.name`
}

class View extends UXBase {
  this.expectedZones = this.props.zoneConfigs.map(...);
  // no constructor
}

class Zone extends UXBase {
  // measureName not needed - use this.props.name
  // onMeasure not needed - use this.props.onMeasure
  // onMark not needed - use this.props.onMark
  // startMarkName not needed, although we still might want it in order to default to 'navigationStart'
  marks = this.props.marks.map(...);
  // no constructor needed unless you want to check for deprecated `config.label`
}
```

Defining that UXBase class and using the `this.props = (object passed into constructor)` pattern would actually save quite a bit of code, and since it matches React's constructor behavior it wouldn't feel too alien to anyone with React experience.

**Comment by sergeychernyshev** [comment]
Does this pattern feel alien to ES6 developers at large? E.g. what if we have a Vue or Angular user here, will they cringe on it or is it OK (even if not default ES6 pattern)?

One thing I am worried about with this pattern is that we lack explicit declaration in the code of which props exist. There is no way to just look at the code and know - @sadafie's approach with all props declared at the top solves this best.

**Comment by mmcgahan** [comment]
It shouldn't feel alien to any developers - IMO it's just a nice simple solution to this kind of problem that the React creators happened to come up with. We can/should definitely document the behavior and the expected property names - type annotations of any flavor will help with that - We could even go so far as to copy the PropTypes interface https://reactjs.org/docs/typechecking-with-proptypes.html

**Comment by sadafie** [comment]
@mmcgahan @sergeychernyshev Updated to use an inherited UXBase class. Ready for final review.

File: js/src/ExpectedMark.js
Comment (mmcgahan) [inline js/src/ExpectedMark.js]: let's skip this 'alias' to keep a single source of 'name' truth

File: js/src/UXBase.js
Comment (mmcgahan) [inline js/src/UXBase.js:12]: code comment to describe intended use of this class?

File: js/src/UXCapture.js
Comment (mmcgahan) [inline js/src/UXCapture.js]: not a blocker, but this comment could also be updated to remove the reference to server-side rendering - there's nothing that inherently ties a View to server rendering

**Review by mmcgahan — COMMENTED** [review]
minor changes requested, but then should be good to go

File: js/src/ExpectedMark.js
Comment (sadafie) [reply js/src/ExpectedMark.js]: `mark.name` is referenced in a few places in - https://github.com/meetup/ux-capture/pull/8/files#diff-1ef3d18dda593c3ce68ba05749eca236R51.

Comment (sadafie) [reply js/src/ExpectedMark.js]: unless you mean that we should use `mark.props.name`? That feels odd to me

Comment (mmcgahan) [reply js/src/ExpectedMark.js]: ah, I see. Interestingly, that oddness is probably the result of some awkward data handling that we can refactor to avoid - essentially we're doing something like

1. get the name of the expected mark
2. create an ExpectedMark with that name
3. listen to the ExpectedMark's `onMark`
4. read the name of the ExpectedMark from the arg to the callback

The name that is returned in (4) is the same name we provided in (1), so we don't actually need it.

Fine to `mark.name` assignment as-is for now, but I might try my hand at making that event handling more streamlined

**Review by mmcgahan — APPROVED** [review]

File: js/src/Zone.js
Comment (mmcgahan) [inline js/src/Zone.js]: can we remove the `this.onMark` and `this.onMeasure` aliases?

File: js/src/ExpectedMark.js
Comment (sadafie) [reply js/src/ExpectedMark.js]: i'll take a stab at this 👍

Comment (sadafie) [reply js/src/ExpectedMark.js]: ah, i think i see how to go about this. it's going to be a slightly bigger diff than i was thinking so i'll do it as a follow on right after merging this, since you also have a ticket blocked by this.

File: js/src/Zone.js
Comment (sadafie) [reply js/src/Zone.js]: yep, removed

**Pull request merged**
